Prius crossing example

pushkalkatara opened the pull request:
Pull request to solve issue #4 

1. Crossing Automata -> Approach Stop Sign + MakeTurn
2. StayonRoad Automata(Maintain Steer and Throttle)
3. Overtaking Automata-> StayonRoad + DetectCarInFront + ChangeLane

pushkalkatara commented:
Hi mentors, @okanasik @jmplaza . The following pull request would hold the new examples as discussed for Prius. Currently the new commit https://github.com/JdeRobot/VisualStates-examples/pull/5/commits/740c53878c41ccdcee443fc9af0562ec2091cb53 holds the crossing automata. Description of the Crossing automata is as follows - 
* Three sub-states -> stay_on_road, stop_on_sign, get_on_highway.
* stay_on_road automata containts P controller to maintain the throttle.
* The node launches the darknet_ros node also which is running Yolo v3 algorithm to detect stop_sign. 
* The transition is a type of Conditional transition, which checkes if the stop_sign is being detected with a probability of 0.5 or greater and based on the prius_front_right_laser_data checks the stop_sign is less than 5 units far from the vehicle.
* The state stop_on_sign activates the breakes and the next transition is a Temporal based transition which makes the vehicle wait for 2 seconds.
* Further state get_on_highway consists two sub-states which are responsible to make a right turn from the vehicle.
Here's a visual example of the above behavior. Please suggest improvements.
https://drive.google.com/open?id=1T5rWmI9mMhB2-JYQXdncQfX98MVI9clK

okanasik commented:
Dear @pushkalkatara,
I have review the pull request. I have changed the dependencies of prius package from gazebo8 to gazebo9. Please, also make sure that in your system you have been testing with gazebo9. Also, I have changed a small part in the README.md file of `prius_example` that clones the darknet.

I am not able to test the whole behavior since `prius_example.launch` file from package `prius_gazebo` is missing. Please add this launch file. Also can you change the name of `prius_example` something more meaningful such as `prius_crossing`. Also we should have standard naming. I think it could be better if we use lowercase + underscore combinations for every directory and example name.

I checked the behavior code. It seems you do not control the steering wheel. My general idea about whole visual processing for prius examples is that. Use of deep network based computer vision algorithms is overkill and can be cumbersome to install and run on a moderate computer. I think we can develop or use (if we can find some open source codes) computer vision algorithms on opencv for sign and lane detection. A simple filter could be enough to detect the road lanes. More importantly, we can change the world to make the perception problem easier. My general view is that we should not focus on the applications that much, we should focus on improving the visualstates itself. Therefore, we can even add a green on the middle of the road to easily detect the lanes. It will not make a big difference, if we solve a tough computer vision problem on simulation as a VisualStates-example. We should find good and easy-to-understand examples that highlights the feature of the tool so that developer will have enough information about the capabilities of the tool for their own application. For example, having prius or kobuki robot and processing laser on them does not make much difference other than showing that it is easy to create a reactive behavior using laser range sensor data. What do you think @jmplaza?

pushkalkatara commented:
Dear Mentors,
I think a discussion over states and transitions would be helpful to improve the example. As suggested by sir @okanasik, a new model specifically for the example would be helpful. I developed the following world in Gazebo. 
![gazebo](https://user-images.githubusercontent.com/21266230/50634211-e6d6c000-0f73-11e9-8ae3-f324b6d8fd6f.gif)
1. Crossing Automata
- In the world, a Human model is going from house to school while crossing the road. 
- The prius detects the stop sign and stop when 4 units near to the laser.
- Detect the human in the region of interest - Road.
- Transition to move when region of interest in clear.
Updates with the following commit - 
- Added function to get Laser data in a specified angle.
- Added function to get Image region of interest (Use of cv_bridge to convert ROS format image to OpenCV compatible matrix)
- Added support for [opencv_apps](https://github.com/ros-perception/opencv_apps/tree/indigo/launch) ROS package for image processing algorithms.
- Added P based throttle controller
Is visual based steer controller required in the example? I think laser based controller would not work in this case, so steer controller based on lane detection would work.

jmplaza commented:
Yes, I agree with your @okanasik: the focus is just on the use of VisualTools more than on developing a robust crossing behavior. Overkilling the problem with complex deeplearning solutions may frighten people to use and run the examples. The point is not to solve the crossing behavior. Instead, it is to use Visualtools to solve it :-) And so, simplifying the problem is advisable, with simple and tuned scenarios.

pushkalkatara commented:
With the recent commit, as discussed the behavior contains the following states - 
1. stay_on_road -> Using prius front camera sensor, image processing technique to detect lanes and proportional controller, prius is following and staying on lanes.
2. stop_on_sign -> Using prius right_laser_sensor, restricting laser angle between 220 and 240 degrees to search for stop sign, and when laser range for stop sign is 9 units, the vehicle stops.
3. human_on_road -> Temporal transition of 6 seconds which waits for the human to cross the road.
4. turn_right (Complex State) -> (4.a) Forward (4.b) SteerRight Makes a turn.
5. align_lane -> Same as stay_on_road and the vehicle aligns itself to the road.
6. Stop -> The behavior ends and the car stops.
![jderobot](https://user-images.githubusercontent.com/21266230/51256878-6cbf2600-19cc-11e9-8fa4-fbfcabfce33c.png)

pushkalkatara commented on prius_crossing/prius_crossing.xml:
```diff
@@ -136,10 +136,8 @@ def calculate_steer_lane(self):
-	error = 1 - abs(slopes[0])
-	p = 0.1 * error
+	slopes = self.get_line_slope() # 0, -1, 1, -2, 2 
+	p = 0
```
Disables Proportional controller for steer following the lanes, as lane detections need to be refined.

pushkalkatara replied:
A better algorithm to select the slope of the line, which is being perfectly detected using Hough Lines. and maintain a threshold distance.

pushkalkatara commented on prius_crossing/prius_crossing.xml:
```diff
@@ -344,7 +342,7 @@ steer = self.namespace.calculate_steer_lane()
 command.throttle = throttle
 command.brake = 0
-command.steer = steer
+command.steer = 0
```
Steer disabled due to line detection.

okanasik commented:
Dear @pushkal, my comments are below;
- For `priusSpawn` example, can you update the README.md file with the latest package organization. For example, there is no package named as `priusData`. The general organization of each example should be just two level. In the current version there is `priusSpawn` directory and in that directory there is another directory for the example. There should a directory with the name of the example and under that directory, there should be behavior xml file. Can you modify the world file launched by `prius_spawn.launch` file. The car should be positioned on top of bridge. It is not easy to see the car when it is positioned under the bridge.

- As a general structure for the examples, it is better not to include the visualstates installation as part of examples. However, you can refer to the installation of visualstates page. In the examples, you refer to the manual install of visualstates, but the user might already have installed the tool from package repository.
 
- It seems, you modify the gazebo worlds by GUI and save that state as world files. This is not a good practice. When you change something through the GUI and save the world file. The world file has the definitions of objects in the world file, but all modifications are saved as additional state in the world file. That makes the world files hard to edit and maintain. Especially, world file for `prius_crossing` seems so hard to maintain. In `prius_crossing` example, I get `Trajectory not found at time` error (I am not sure whether that is related to the actor, but it may also refer to some missing world objects) Tomorrow, I will edit and clear the world files. Once I clear up the world file, I will review the behavior. Best.

okanasik commented:
Dear @pushkalkatara, I pushed the updated world files. If you need to change the position of any object please open the world file with a text editor and change the `<pose>` element. Best,

okanasik commented:
Dear @pushkalkatara, I updated the world of prius crossing example. I use simple thresholding to differentiate the road from the things around the car. The car detects the right line and calculate the slope of the right line and tries to keep at -1.0 radian angle for that line. It is not perfect, but it good enough for that simple example. Can you test it?

pushkalkatara commented:
Dear @okanasik , I checked with the slope implementation, it returned integer values like -1,0,1 only which were not syncing with the P controller. I rather used the distance between the center of the camera and the detected left lane contour in the example. The P controller is maintaining this distance and following the road.

okanasik commented:
- On my system I having the following error:
```
File "/media/okan/depo/jderobot/catkin_ws/src/VisualStates-examples/prius_crossing/prius_crossing.py", line 234, in runCode
    steer = self.namespace.calculate_steer_lane()
  File "/media/okan/depo/jderobot/catkin_ws/src/VisualStates-examples/prius_crossing/prius_crossing.py", line 138, in calculate_steer_lane
    error = self.get_steer_error()
  File "/media/okan/depo/jderobot/catkin_ws/src/VisualStates-examples/prius_crossing/prius_crossing.py", line 173, in get_steer_error
    M = cv2.moments(contours[0])
IndexError: list index out of range
```
This error occurs when the car hits the speed limit sign. I think there is an error about the timing of turn behavior that can occur differently on different systems. If you fix the error above, I plan to merge and improve incrementally.
- Can you remove the robot_arm since it is already merged to master with other pull request.
- Can you also remove the <state> </state> tag from the world file. If you remember, we have discussed that the world files should be human readable and editable.
- At the end, just before the merge, can you update the README.md file with the latest state of the example. such as updating the directory names.

pushkalkatara commented:
Demo Video for pull-request - https://www.youtube.com/watch?v=Tqh14Q2boXY

okanasik commented:
The world seen in video is different on my system. You can see the gazebo scene below. The house is just on the road. Also it is a little weird that human is hitting the building after crossing :)

![gazebo_crossing](https://user-images.githubusercontent.com/1243277/54287394-9e143680-45b6-11e9-90d1-615f3b465093.png)

Also the visual states screen is different than the one on the video. You can see below the current behavior. Unfortunately, the behavior does not work on my system as expected, but once we clear these two issues we can merge.
![visualstates](https://user-images.githubusercontent.com/1243277/54287645-1b3fab80-45b7-11e9-8df7-9aeb66240718.png)

I also made small change since those part of the code was giving run-time errors.

pushkalkatara commented:
Hi @okanasik sir, I am not able to reproduce why the gazebo model is being rendered differently on your system, I crossed checked the environment recloning the example. Also, the prius_crossing.xml is added with the TurnRight automata too, weird if only 3 states are shown. 
I had the changes following this change https://github.com/JdeRobot/VisualStates-examples/pull/5/commits/8e9fa1c4ee73b1986a3b558454bd9edfd87c0399 .
